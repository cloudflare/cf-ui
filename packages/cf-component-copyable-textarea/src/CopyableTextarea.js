const React = require('react');
const { PropTypes } = React;
const Textarea = require('cf-component-textarea');

class CopyableTextarea extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      helpText: this.props.clickToCopyText,
      copied: false
    };

    this.handleFocus = this.handleFocus.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
  }

  handleFocus(e) {
    e.target.select();

    const { onCopy } = this.props;

    let success;
    try {
      success = document.execCommand('copy');
    } catch (err) {
      success = false;
    }

    if (success && onCopy) {
      onCopy();
    }

    this.setState({
      helpText: success
        ? this.props.copiedTextToClipboardText
        : this.props.pressCommandOrCtrlCToCopyText,
      copied: success
    });
  }

  handleBlur() {
    this.setState({
      helpText: this.props.clickToCopyText,
      copied: false
    });
  }

  render() {
    return (
      <div className="cf-copyable-textarea">
        <Textarea
          ref={node => this.textarea = node}
          readOnly
          name={this.props.name}
          value={this.props.value}
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
        />
        <p className="cf-copyable-textarea__help-text">{this.state.helpText}</p>
      </div>
    );
  }
}

CopyableTextarea.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  onCopy: PropTypes.func,
  clickToCopyText: PropTypes.string,
  copiedTextToClipboardText: PropTypes.string,
  pressCommandOrCtrlCToCopyText: PropTypes.string
};

CopyableTextarea.defaultProps = {
  clickToCopyText: 'Click to copy',
  copiedTextToClipboardText: 'Copied text to clipboard',
  pressCommandOrCtrlCToCopyText: 'Press Command/Ctrl+C to copy'
};

module.exports = CopyableTextarea;
