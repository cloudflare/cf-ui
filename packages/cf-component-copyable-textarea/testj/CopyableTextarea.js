import React from 'react';
import renderer from 'react-test-renderer';
import jsdom from 'jsdom';
import { mount } from 'enzyme';
import CopyableTextarea from 'cf-component-copyable-textarea';

beforeEach(() => {
  const doc = jsdom.jsdom('<!doctype html><html><body></body></html>');
  global.document = doc;
  global.window = doc.defaultView;
  global.document.execCommand = () => true;
});

test('should render', () => {
  const component = renderer.create(
    <CopyableTextarea
      name="example"
      value="content"
      clickToCopyText="click here"
      copiedTextToClipboardText="copied"
      pressCommandOrCtrlCToCopyText="press here"
    />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('should call onCopy on focus', () => {
  let clicked = false;
  const copyableTextarea = mount(
    <CopyableTextarea
      name="example"
      value="content"
      onCopy={() => clicked = true}
    />
  );
  copyableTextarea.find('textarea').simulate('focus');
  expect(clicked).toBeTruthy();
});

test('should render copied help text after copying', () => {
  const copyableTextarea = mount(
    <CopyableTextarea
      name="example"
      value="content"
      copiedTextToClipboardText="copied"
    />
  );
  copyableTextarea.find('textarea').simulate('focus');
  expect(copyableTextarea.find('.cf-copyable-textarea__help-text').text()).toBe(
    'copied'
  );
});

test('should render press help text if copying was not successful', () => {
  global.document.execCommand = () => {
    throw new Error('error');
  };
  const copyableTextarea = mount(
    <CopyableTextarea
      name="example"
      value="content"
      pressCommandOrCtrlCToCopyText="press here"
    />
  );
  copyableTextarea.find('textarea').simulate('focus');
  expect(copyableTextarea.find('.cf-copyable-textarea__help-text').text()).toBe(
    'press here'
  );
});
