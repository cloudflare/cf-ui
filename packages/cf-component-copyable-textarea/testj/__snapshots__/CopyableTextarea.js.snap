// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="cf-copyable-textarea"
>
  <textarea
    className="cf-textarea cf-textarea--readonly"
    name="example"
    onBlur={[Function]}
    onFocus={[Function]}
    readOnly={true}
    value="content"
  />
  <p
    className="cf-copyable-textarea__help-text"
  >
    click here
  </p>
</div>
`;
